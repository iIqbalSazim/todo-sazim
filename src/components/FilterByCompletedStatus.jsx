import { Tabs } from "@mantine/core";
import { COLORS } from "../constants/constant";

const FilterByCompletedStatus = ({ setFilter }) => {
  return (
    <Tabs
      orientation="vertical"
      variant="pills"
      color={COLORS.BUTTON}
      defaultValue="all"
      onTabChange={(value) => setFilter(value)}
      h={"180px"}
    >
      <Tabs.List position="apart">
        <Tabs.Tab value="all">All</Tabs.Tab>
        <Tabs.Tab value="active">Active</Tabs.Tab>
        <Tabs.Tab value="completed">Completed</Tabs.Tab>
      </Tabs.List>
    </Tabs>
  );
};

export default FilterByCompletedStatus;
