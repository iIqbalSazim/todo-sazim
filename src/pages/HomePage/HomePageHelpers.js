export function findIndexWithId(data, id) {
  return data.findIndex((element) => element.id === id);
}

export function formatCompletedAt(dateString) {
  const date = new Date(dateString);

  let currentTime =
    date.toLocaleTimeString().slice(0, 4) + date.toLocaleTimeString().slice(7);
  let currentDate = `${currentTime} | ${date.toDateString()}`;
  return currentDate;
}

export function formatDueDate(dateString) {
  const date = new Date(dateString);
  return date.toDateString().slice(0, 15);
}

export const replaceTabId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-tab-(all|active|completed|target)"/g,
    (match, p1) => `id="mantine-mockedId-tab-${p1}"`
  );
};

export const replaceMenuTargetId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-(target)"/g,
    (match, p1) => `id="mantine-mockedId-${p1}"`
  );
};

export const replaceDropdownTargetId = (htmlString) => {
  return htmlString.replace(
    /id="mantine-[a-z0-9]+-(dropdown)"/g,
    (match, p1) => `id="mantine-mockedId-${p1}"`
  );
};

export const replaceDataFocusId = (htmlString) => {
  return htmlString.replace(
    /data-focus-id="mantine-[a-z0-9]+"/g,
    'data-focus-id="mantine-mockedId"'
  );
};

export const replaceAriaLabelledBy = (htmlString) => {
  return htmlString.replace(
    /aria-labelledby="mantine-[a-z0-9]+-target"/g,
    'aria-labelledby="mantine-mockedId-target"'
  );
};

export const replaceAriaControls = (htmlString) => {
  return htmlString.replace(
    /aria-controls="mantine-[a-z0-9]+-dropdown"/g,
    'aria-controls="mantine-mockedId-dropdown"'
  );
};

export const replaceAllRandomGeneratedIds = (htmlString) => {
  let result = replaceTabId(htmlString);
  result = replaceMenuTargetId(result);
  result = replaceDropdownTargetId(result);
  result = replaceDataFocusId(result);
  result = replaceAriaLabelledBy(result);
  result = replaceAriaControls(result);

  return result;
};
