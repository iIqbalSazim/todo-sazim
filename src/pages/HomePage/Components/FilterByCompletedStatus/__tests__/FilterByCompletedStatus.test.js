import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { replaceAllRandomGeneratedIds } from "../../../HomePageHelpers";
import FilterByCompletedStatus from "../FilterByCompletedStatus";

const mockSetFilter = jest.fn();

describe("FilterByCompletedStatus", () => {
  test("renders correctly", () => {
    const { container } = render(
      <FilterByCompletedStatus setFilter={mockSetFilter} />
    );

    const allTab = screen.getByRole("tab", {
      name: /all/i,
    });
    const activeTab = screen.getByRole("tab", {
      name: /active/i,
    });
    const completedTab = screen.getByRole("tab", {
      name: /completed/i,
    });

    expect(allTab).toBeInTheDocument();
    expect(activeTab).toBeInTheDocument();
    expect(completedTab).toBeInTheDocument();

    expect(replaceAllRandomGeneratedIds(container.innerHTML)).toMatchSnapshot();
  });

  test("calls setFilter with value 'all' when clicked on All tab", async () => {
    const { container } = render(
      <FilterByCompletedStatus setFilter={mockSetFilter} />
    );

    const allTab = screen.getByRole("tab", {
      name: /all/i,
    });

    await userEvent.click(allTab);

    expect(mockSetFilter).toBeCalledWith("all");

    expect(replaceAllRandomGeneratedIds(container.innerHTML)).toMatchSnapshot();
  });

  test("calls setFilter with value 'active' when clicked on Active tab", async () => {
    const { container } = render(
      <FilterByCompletedStatus setFilter={mockSetFilter} />
    );

    const activeTab = screen.getByRole("tab", {
      name: /active/i,
    });

    await userEvent.click(activeTab);

    expect(mockSetFilter).toBeCalledWith("active");
    expect(replaceAllRandomGeneratedIds(container.innerHTML)).toMatchSnapshot();
  });

  test("calls setFilter with value 'completed' when clicked on Completed tab", async () => {
    const { container } = render(
      <FilterByCompletedStatus setFilter={mockSetFilter} />
    );

    const completedTab = screen.getByRole("tab", {
      name: /completed/i,
    });

    await userEvent.click(completedTab);

    expect(mockSetFilter).toBeCalledWith("completed");
    expect(replaceAllRandomGeneratedIds(container.innerHTML)).toMatchSnapshot();
  });
});
