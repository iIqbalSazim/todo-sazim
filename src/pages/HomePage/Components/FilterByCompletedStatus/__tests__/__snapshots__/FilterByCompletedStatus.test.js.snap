// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterByCompletedStatus calls setFilter with value 'active' when clicked on Active tab 1`] = `"<div class="mantine-Tabs-root mantine-o79pz9"><div class="mantine-Tabs-tabsList mantine-1dphvmd" role="tablist" aria-orientation="vertical"><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-all" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">All</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-active" aria-selected="true" tabindex="0" data-active="true"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Active</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-completed" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Completed</span></button></div></div>"`;

exports[`FilterByCompletedStatus calls setFilter with value 'all' when clicked on All tab 1`] = `"<div class="mantine-Tabs-root mantine-o79pz9"><div class="mantine-Tabs-tabsList mantine-1dphvmd" role="tablist" aria-orientation="vertical"><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" data-active="true" role="tab" id="mantine-mockedId-tab-all" aria-selected="true" tabindex="0"><span class="mantine-480qv8 mantine-Tabs-tabLabel">All</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-active" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Active</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-completed" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Completed</span></button></div></div>"`;

exports[`FilterByCompletedStatus calls setFilter with value 'completed' when clicked on Completed tab 1`] = `"<div class="mantine-Tabs-root mantine-o79pz9"><div class="mantine-Tabs-tabsList mantine-1dphvmd" role="tablist" aria-orientation="vertical"><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-all" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">All</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-active" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Active</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-completed" aria-selected="true" tabindex="0" data-active="true"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Completed</span></button></div></div>"`;

exports[`FilterByCompletedStatus renders correctly 1`] = `"<div class="mantine-Tabs-root mantine-o79pz9"><div class="mantine-Tabs-tabsList mantine-1dphvmd" role="tablist" aria-orientation="vertical"><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" data-active="true" role="tab" id="mantine-mockedId-tab-all" aria-selected="true" tabindex="0"><span class="mantine-480qv8 mantine-Tabs-tabLabel">All</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-active" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Active</span></button><button class="mantine-UnstyledButton-root mantine-Tabs-tab mantine-1l1luj0" type="button" role="tab" id="mantine-mockedId-tab-completed" aria-selected="false" tabindex="-1"><span class="mantine-480qv8 mantine-Tabs-tabLabel">Completed</span></button></div></div>"`;
